Dockerfile

```diff
@@ -334,28 +326,57 @@ ENV LD_LIBRARY_PATH="/usr/local/lib:/usr/lib" \
     CXXFLAGS="" \
     NUMTHREADS=$NUMTHREADS
 
+FROM build_grass_config AS build_grass_with_gui
+
+ARG GRASS_CONFIG="$GRASS_CONFIG $GRASS_GUI_CONFIG"
+ARG GRASS_GUI_PACKAGES="$GRASS_GUI_RUN_PACKAGES $GRASS_GUI_BUILD_PACKAGES"
+
+# hadolint ignore=DL3008
+RUN echo "Installing GRASS GUI packages: $GRASS_GUI_PACKAGES" \
+    && apt-get update \
+    && apt-get upgrade -y \
+    && apt-get install -y --no-install-recommends --no-install-suggests \
+    $GRASS_GUI_PACKAGES \
+    && apt-get autoremove -y \
+    && apt-get clean all \
+    && rm -rf /var/lib/apt/lists/* \
+    # Need to generate locale when NLS is enabled
+    && echo LANG="en_US.UTF-8" > /etc/default/locale \
+    && echo en_US.UTF-8 UTF-8 >> /etc/locale.gen \
+    && locale-gen
+
+FROM build_grass_config AS build_grass_without_gui
+
+ARG GRASS_CONFIG="$GRASS_CONFIG $GRASS_NO_GUI_CONFIG"
+
+# hadolint ignore=DL3006
+FROM build_grass_${GUI}_gui AS build_grass
+
 # Configure compile and install GRASS
-# hadolint ignore=SC2086
+# hadolint ignore=SC2086,DL3008
 RUN make -j $NUMTHREADS distclean || echo "nothing to clean" \
+    && make -j $NUMTHREADS distclean || echo "nothing to clean" \
     && ./configure $GRASS_CONFIG \
-    $GRASS_ADDITIONAL_CONFIG \
     && make -j $NUMTHREADS \
     && make install && ldconfig \
     && rm -rf /usr/local/grass85/demolocation \
-    && if [ "$GUI" = "with" ] ; then \
-        echo "GUI selected, skipping GUI related cleanup"; \
-    else \
-        echo "No GUI selected, removing GUI related files"; \
-        cp /usr/local/grass85/gui/wxpython/xml/module_items.xml module_items.xml \
-        && rm -rf /usr/local/grass85/fonts \
-        && rm -rf /usr/local/grass85/gui \
-        && rm -rf /usr/local/grass85/docs/html \
-        && rm -rf /usr/local/grass85/docs/mkdocs \
-        && rm -rf /usr/local/grass85/share \
-        && mkdir -p /usr/local/grass85/gui/wxpython/xml/ \
-        && mv module_items.xml /usr/local/grass85/gui/wxpython/xml/module_items.xml; \
-    fi
+    && cp /usr/local/grass85/gui/wxpython/xml/module_items.xml module_items.xml \
+    && mkdir -p /usr/local/grass85/gui/wxpython/xml/ \
+    && mv module_items.xml /usr/local/grass85/gui/wxpython/xml/module_items.xml
+
+FROM build_grass AS build_grass_with_gui_built
+RUN echo "GUI selected, skipping GUI related cleanup"
+
+FROM build_grass AS build_grass_without_gui_built
+RUN echo "No GUI selected, removing GUI related files" \
+    && rm -rf /usr/local/grass85/fonts \
+    && rm -rf /usr/local/grass85/gui \
+    && rm -rf /usr/local/grass85/docs/html \
+    && rm -rf /usr/local/grass85/docs/mkdocs \
+    && rm -rf /usr/local/grass85/share
```

Comment:
Note: Removing files in a separate run step only works because we are copying files from that stage into the final stage (or a stage that the final stage is based on). It wouldn't reduce size if it was a stage that had layers uploaded.

Also, this means that the layer that has these extra files is uploaded to cache with cache mode max that we currently use.


Was the grass build stage needed to be the same for both variants to be useful?